### PR TITLE
Feat/add trip details

### PR DESCRIPTION
**Pull request by alvynben**
# PR Summary
This PR introduces enhancements to the way we handle and present trip details, and it also adds utility functions to facilitate date and time operations.

# Description

### Trip Detail Query Enhancement:

The detail query function fetches a specific trip using its id. If the trip is found and contains events, it further refines the details.
Events associated with the trip are fetched (potentially in a single database call for optimized performance).
The events are sorted chronologically using the getSecondsBetweenTwoTimestamps helper.
The duration of the trip is broken down into days, and events are categorized accordingly.
The function finally returns a structured trip detail object, including the trip's name and its daily events.
(Reference: detail function)

### Date Utilities:

getDates: This utility function returns an array of dates (along with their start and end times) between a specified startTime and endTime. This is particularly beneficial for mapping out events within a trip's duration.
getSecondsBetweenTwoTimestamps: A function to quickly determine the difference in seconds between two timestamps, aiding in sorting and time comparisons.
(Reference: getDates & getSecondsBetweenTwoTimestamps functions)

### CODE REVIEW

File: aitinero/convex/utils.ts [hunk 1]
```diff
@@ -1,10 +1,33 @@
...
+      day: newDate.toLocaleString(),
+      start: newDate.toISO()!,
+      end: newDate.endOf('day').toISO()!,
+    });
+    currentDate = currentDate.plus({ days: 1 });
+  }
+  return dateArray;
+}
+
+export const getSecondsBetweenTwoTimestamps = (startTime: string, endTime: string) => {
```
Comment (alvynben) [inline aitinero/convex/utils.ts:25]: Do better naming.
```suggestion
export const getSecondsBetweenTimestamps = (startTime: string, endTime: string) => {
```

File: aitinero/convex/trips.ts [hunk 2]
```diff
@@ -63,3 +64,32 @@ export const list = query({
...
+        let events = await Promise.all(trip.events.map(ctx.db.get));
+
+        // If events are always in chronological order in the DB, this sort might not be necessary.
+        events.sort((event1, event2) => getSecondsBetweenTwoTimestamps(event2!.start_time, event1!.start_time));
+
+        const days = getDates(trip.start_date, trip.end_date);
+        const tripDetails = days.map(day => ({
+            date: day.day,
+            events: events.filter(event => event!.start_time > day.start && event!.end_time < day.end)
+        }));
```
Comment (alvynben) [inline aitinero/convex/trips.ts:88]: Need to improve logic here:
* Need to account for events split within days 
* Could improve performance (stop looking at all events for each day) [Low priority - still O(1) currently with data size]

File: aitinero/convex/trips.ts [hunk 3]
```diff
@@ -63,3 +64,32 @@ export const list = query({
...
+    handler: async (ctx, { }) => {
+        const trip = await ctx.db.query(TABLE_NAME.TRIPS).unique();
+
+        if (!trip || !trip.events || trip.events.length === 0) {
+            return {};
+        }
+
+        // Assuming getAll fetches events based on event_ids. 
+        // It's beneficial if the underlying method has a way of fetching these in a single DB call.
+        let events = await Promise.all(trip.events.map(ctx.db.get));
```
Comment (alvynben) [inline aitinero/convex/trips.ts:79]: Need to do some accounting to make sure events are not null.

File: aitinero/convex/utils.ts [hunk 4]
```diff
@@ -1,10 +1,33 @@
-import { DateTime } from 'luxon';
+import { GenericId } from 'convex/values';
+import { DateTime, DiffOptions, DurationUnits } from 'luxon';
 
 export const convertTimestampToMilliseconds = (timestamp: string) => {
   const date = DateTime.fromISO(timestamp);
   return date.toMillis();
 };
 
+export const getDates = (startTime: string, endTime: string) => {
```
Comment (alvynben) [inline aitinero/convex/utils.ts:9]: Got a better naming convention?

**Review by alvynben — COMMENTED** [review]
- [x] Changes made?

File: aitinero/convex/utils.ts (hunk 1 above)
Comment (alvynben) [reply aitinero/convex/utils.ts:25]: done!

File: aitinero/convex/trips.ts (hunk 2 above)
Comment (alvynben) [reply aitinero/convex/trips.ts:88]: # Changes

## Event Fetching:

Original: The events were fetched and directly sorted.
New: After fetching the events, there's an added step to filter out any null or undefined events using `events.filter(event => !!event);`. This ensures that only valid events are considered in subsequent operations.

## Event Filtering within Days:

Original: The filter condition for events within days was based on both the start_time and end_time of events.
New: The filter condition is now solely based on the start_time of events with the condition `event!.start_time > day.start && event!.start_time < day.end`.

File: aitinero/convex/trips.ts (hunk 3 above)
Comment (alvynben) [reply aitinero/convex/trips.ts:79]: done!

**Review by xav168 — APPROVED** [review]
lgtm

**Pull request merged**